### ISSUE TITLE
[Idea] Reorganising status code

### ISSUE THREAD

**Issue by kylechui** [andweeb/presence.nvim#13]
I was doing some more thinking yesterday, and thought of adding new variables like `reading_text` and `browsing_text` to complement the existing `editing_text`. Then the `get_status_text` function could just return the corresponding variable depending on context, rather than repeatedly set `editing_text`. From my POV this has two main benefits:
1) More customisability for the end user (as they can set more variables for more contexts)
2) No need to deprecate `editing_text`, nor change any of the existing logic for the `get_status_text` function

Let me know what you think, and I'll start working on a test version of this code

**Comment by andweeb** [comment]
Good idea! I was thinking along the same lines too. Yeah I agree, the current `editing_text` option is a little too generic (and a bit of a misnomer) and requires users to reimplement these various checks to configure the text, so having these separate options to configure different texts at different states would be 💯.

What do you think about the following options for the following states?
* `editing_text` for when the buffer is modifiable and not read-only
* `reading_text` for when the buffer is not modifiable or read-only
* `browsing_text` for when the buffer is a supported file tree (maybe `filetree_text` or `navigating_text`? not sure if "browsing" would denote "browsing/navigating file tree")
* `committing_text` for when the buffer is a git commit (or `gitcommit_text`?)

We can implement these status text options for now as a foundational step, and eventually this can open up more statuses like your idea in your [other comment](https://github.com/andweeb/presence.nvim/issues/9#issuecomment-842878172), something like:
* `idling_text` for when the user is idling/inactive based on `CursorMoved`
* `creating_text` for when the user is on a new unnamed buffer (**TBD** on this one)

**Comment by kylechui** [comment]
I have most(?) of the rewrite done on the `dynamic_text_update` branch of my [fork](https://github.com/kylechui/presence.nvim/tree/dynamic_text_update), but for some reason I keep running into the issue `options cannot be indexed because it is nil`, which leads me to suspect that the `setup` function isn't being called properly. I can't seem to figure out how to make this last bit work properly

**Comment by andweeb** [comment]
Oh wow nice! Okay, I'll check it out later since I'm a bit busy at the moment, I'll get back to you sometime tonight!

**Comment by andweeb** [comment]
Hi @kylechui! I found the issue :) Seems like we need to update that `set_option` call. Here's an annotated pseudocode snippet of the relevant changes:
```lua
function Presence:setup(options)
    -- ... setup code...

    -- Passing in `self.get_status_text` function as the default option value
    self:set_option("status_text", self.get_status_text)

    -- ... more setup code...
end

function Presence:update_for_buffer(args)
    -- ... code...

    -- When using the default status text option value, this line will invoke the
    -- `self.get_status_text` function like this: `self.get_status_text(filename, buffer)`
    local status_text = self.options.status_text(filename, buffer)

    -- ... more code...
end
```

Note how the `get_status_text` method is currently being invoked via dot syntax, when instead we should be invoking via colon syntax, like `self:get_status_text(...)`. This results in that nil error because the implicit first argument for `get_status_text` expects the `Presence` module (as `self`), but since it's the `filename` string variable instead due to the dot syntax, it tries to index the option keys in `filename.options` which of course is `nil`.

You can read about the distinction between dot vs colon syntax in Lua [here](https://www.lua.org/pil/16.html) if you're unfamiliar!

So in other words, we need to be able to pass the `get_status_text` function with the correct `self` argument when passing it to `set_options` in the `setup` function. We can do that by creating an inline wrapper function that properly invokes `get_status_text` using the colon syntax with a pass-through of all args.

**tldr;** You can fix the error by setting the `status_text` option like this instead:
```lua
self:set_option("status_text", function(...)
    return self:get_status_text(...)
end)
```

I hope that's clear, let me know if you have any other questions!

**Comment by kylechui** [comment]
Thanks for taking the time to help me out with this stuff! I just patched it and made sure the status texts were working as intended (although not with `vim-plug` because I don't use that, although patching it would be an easy fix later down the line if it is wrong).

**Comment by andweeb** [comment]
Closing as this was achieved in #15 :)
